### ISSUE TITLE
multipart mime with 2 HTML Sections

### ISSUE THREAD

**Issue by kalzon1** [bertjohnson/OpaqueMail#62]
I appreciate the work you do to keep this up to date.

I've been receiving a few multipart mime files where there are 2 Content-Type: text/html sections

The code in 
MailMessages.ProcessMimeParts()

approx line1277 (See below)
overwrites the 1st HTML section with the second.  I don't know why these email are coming from people with this double HTML section, but they are.  So in testing it, I think it needs to be more something like

                        if (Body.ToUpper().StartsWith("<HTML"))
                        {
                            Body = Body + mimePart.Body;
                        }
                        else
                        {
                            Body = mimePart.Body;
                        }

I have some sample emails I could send you to test yourself if you'd like, and a better solution than this would be great too.

Thank you.

                else
                {
                    // If the current body isn't text/html and this is, replace the default body with the current MIME part.
                    if (!ContentType.ToUpper().StartsWith("TEXT/HTML") && contentTypeToUpper.StartsWith("TEXT/HTML"))
                    {
                        // Add the previous default body as an alternate view.
                        MemoryStream alternateViewStream = new MemoryStream(Encoding.UTF8.GetBytes(Body));
                        AlternateView alternateView = new AlternateView(alternateViewStream, BodyContentType);
                        if (BodyTransferEncoding != TransferEncoding.Unknown)
                            alternateView.TransferEncoding = BodyTransferEncoding;
                        AlternateViews.Add(alternateView);

                        IsBodyHtml = true;
                        Body = mimePart.Body;
                        BodyContentType = mimePart.ContentType;
                        CharSet = mimePart.CharSet;
                    }
                    else
                    {
                        // If the MIME part isn't of type text/*, treat is as an attachment.
                        MemoryStream attachmentStream = new MemoryStream(mimePart.BodyBytes);
                        Attachment attachment;
                        if (mimePart.ContentType.IndexOf("/") > -1)
                            attachment = new Attachment(attachmentStream, mimePart.Name, mimePart.ContentType);
                        else
                            attachment = new Attachment(attachmentStream, mimePart.Name);
                        attachment.ContentId = mimePart.ContentID;
                        Attachments.Add(attachment);
                    }
                }

**Comment by bertjohnson** [comment]
Thanks for the report, kalzon1. Odd that you're receiving email with multiple text/html MIME parts.

The specs state that the last MIME part that can be rendered should be considered the highest fidelity, so only that one part should be considered our body. I see that the current code is greedy, sticking with the first text/html part it finds. I'll review and update that.

For troubleshooting, would you be able to email me an example? My email is bert AT (this project name) DOT org.

**Comment by bertjohnson** [comment]
Thanks again for the report.

I just uploaded release 2.5.3. When there are multiple text/html parts, the last one will be the body. Previous text/html parts will be alternative views. Please test it out.

**Comment by kalzon1** [comment]
Hi Bert,


I sent it to you in 2 ways.  Hopefully one of them will not have 'cleaned it up' somehow.  Please let me know if it does not come through like that.  I can try and send it a different way.


In the example I emailed you the 1st HTML section is the actual message, the second is really nothing.  I don't know why there are two or if sometimes it might be the 2nd one that they want, which is why I coded with an appending rather than an alternate, because I don't know how I would offer the user to look at an alternate.  What are your thoughts.


Thanks,

Keith Longwell


________________________________

**Comment by kalzon1** [comment]
I know I'm repeating but it better matched this email.


In the case of the emails I have as samples from real emails, it is the 1st text/html part that is the real email.  If it had been the last your code would have worked because it replaced it with the last.


So if you offer an alternative view, how might you suggest I would show that to a user?


Thanks

Keith Longwell


________________________________

**Comment by jstedfast** [comment]
Are the text/html parts part of a multipart/mixed or a multipart/alternative?

If they are part of a multipart/mixed, standard practice is to use the first text/html part as the body. For multipart/alternatives, it the last part (i.e. they are in the order of least-faithful to most faithful renderability of the message as composed by the sender).

*Sometimes* what happens is that mailing-list software decides to munge messages by appending its own footer and *sometimes* these mailing-list programs are dumb when it comes to multipart/mixed vs multipart/alternative and so they just blindly append their own text/plain or text/html part(s) into whatever multipart they find.

**Comment by kalzon1** [comment]
it is a multipart/mixed.


The 1st text/html part is part of the body, and the 2nd is "other stuff".


But the code, as it was, was using the LAST text/html part as the actual body.


Should the code just be changed so that if the 1st text/html was already loaded, to not overwrite it but leave that as the main/real body if multippart/mixed?



________________________________

**Comment by jstedfast** [comment]
Some mail clients will render inline text parts back-to-back as if they were combined.

You could do that.

As far as your last question: yes, the first text/html part should be considered the main "body" - at least if you consider the body to be a single part. Other text/html parts should be considered supplemental to the first.

**Comment by bertjohnson** [comment]
kalzon1, thanks for the sample emails. And jstedfast, thanks for weighing in on the multipart/mixed vs. multipart/alternative distinction. I need to revisit the specs.

I'll take another look and update in the next few days. It sounds like the first text/html should be considered the body in some (if not all) scenarios, so I'll make sure that works.

**Comment by jstedfast** [comment]
@bertjohnson correct, the first text/html part should be considered the body *except* when children of a multipart/alternative (and by children-of, I don't necessarily mean *direct* children of).

As an illustration of the types of multipart/alternative structures that you are *likely* to encounter, they look like this:

The simple case:
```
multipart/alternative
    text/plain
    text/html
```

When the html part has related images:
```
multipart/alternative
    text/plain
    multipart/related
        text/html
        image/jpeg
        ...
```
-or-
```
multipart/related
    multipart/alternative
        text/plain
        text/html
    image/jpeg
    ...
```

Keep in mind that multipart/related parts are a bit more complicated than simply assuming that the first part is the root document (altho it generally is). You'll need to check the `start` parameter on the multipart/related's Content-Type header - if unset, then it's the first part, otherwise it contains the Content-Id value of the root document.

Of course, I've also seen multipart/mixed's inside of multipart/alternatives and all sorts of other fun stuff, but they are rare.

**Comment by bertjohnson** [comment]
That's a great explanation. Thanks for taking the time, Jeffrey. I'll update OpaqueMail when I get a chance.

**Comment by jstedfast** [comment]
No prob!

**Comment by bertjohnson** [comment]
I think that the logic is corrected as of release 2.5.4. kalzon1, it mirrors what you shared in your original post. Thanks again.

By the way, I've added a note to the "README.md" warning that this project isn't actively maintained. I'll still review and respond when I can, but I realize these month-long hiatuses aren't ideal. MailKit is better for most use cases and Jeffrey's great at responding, so I'll point users towards that project.
